### PR TITLE
ids[multipart]: add ids multipart handler for contract messages

### PR DESCRIPTION
**Pull request by juliapampus**
Depends on PR #284

### CODE REVIEW

**Comment by juliapampus** [comment]
@denisneuling @DominikPinsel Small problem: In IDS, there are also messages for which no response is expected, e.g. a NotificationMessage. I would send a MessageProcessedNotificationMessage to an ContractAgreement and RequestInProcessMessages to Contract Offers. I wanted to add a handler for both message types that just log the incoming message and do nothing. However, I cannot return null in the handleRequest method as the MultipartController would then automatically send a RejectionMessage. Already thought about what can be done to prevent this?

**Comment by denisneuling** [comment]
@juliapampus Not sure whether I understood correctly, but would it help to make `MultipartResponse#header` a nullable field and return `MultipartResponse.Builder.newInstance().build();` in your handler? If so, this would lead to an successful but empty multipart response. - Analoguous we would have to adapt the dispatcher module making `IdsMultipartParts#header` nullable as well.

**Comment by tmberthold** [comment]
(To compare the current state, the Messaging-Services in this case simply [return a status code 200](https://github.com/International-Data-Spaces-Association/IDS-Messaging-Services/blob/aa36a2dbf99cbefc202692c050eb092cf905935f/messaging/src/main/java/de/fraunhofer/ids/messaging/endpoint/MessageController.java#L192) if the handler does return null.)

**Comment by juliapampus** [comment]
`CosmosContractNegotiationStoreTest` still failing - maybe you could take a quick look @paullatzelsperger or @DominikPinsel ? Apart from that, this is ready for review. Tests could be improved during the Hackathon (new PR) @jimmarino

File: data-protocols/ids/ids-api-multipart-endpoint-v1/src/main/java/org/eclipse/dataspaceconnector/ids/api/multipart/handler/ArtifactRequestHandler.java [hunk 1]
```diff
@@ -118,11 +120,11 @@ public boolean canHandle(@NotNull MultipartRequest multipartRequest) {
...
                 .policy(contractDefinition.get().getContractPolicy())
-                .contractEndDate(Instant.now().getEpochSecond() + 60 * 5 /* Five Minutes */)
-                .contractSigningDate(Instant.now().getEpochSecond() - 60 * 5 /* Five Minutes */)
-                .contractStartDate(Instant.now().getEpochSecond() - 60 * 5 /* Five Minutes */)
-                .consumerAgentId("https://example.com")
-                .providerAgentId("https://example.com")
+                .contractEndDate(ZonedDateTime.ofInstant(Instant.ofEpochMilli(Instant.now().getEpochSecond() + 60 * 5), ZoneId.of("UTC")) /* Five Minutes */)
+                .contractSigningDate(ZonedDateTime.ofInstant(Instant.ofEpochMilli(Instant.now().getEpochSecond() - 60 * 5), ZoneId.of("UTC")) /* Five Minutes */)
+                .contractStartDate(ZonedDateTime.ofInstant(Instant.ofEpochMilli(Instant.now().getEpochSecond() - 60 * 5), ZoneId.of("UTC")) /* Five Minutes */)
+                .consumerAgentId(URI.create("https://example.com"))
```
[the remark this answers is not in the document]
Comment (juliapampus) [reply data-protocols/ids/ids-api-multipart-endpoint-v1/src/main/java/org/eclipse/dataspaceconnector/ids/api/multipart/handler/ArtifactRequestHandler.java]: I think @DominikPinsel already has this on his todo list

File: spi/src/main/java/org/eclipse/dataspaceconnector/spi/types/domain/contract/agreement/ContractAgreement.java [hunk 2]
```diff
@@ -30,20 +32,20 @@
...
 
     private final String id;
-    private final String providerAgentId;
-    private final String consumerAgentId;
-    private final long contractSigningDate;
-    private final long contractStartDate;
-    private final long contractEndDate;
+    private final URI providerAgentId; // TODO change to string again?
+    private final URI consumerAgentId;
+    private final ZonedDateTime contractSigningDate;
```
[the remark this answers is not in the document]
Comment (juliapampus) [reply spi/src/main/java/org/eclipse/dataspaceconnector/spi/types/domain/contract/agreement/ContractAgreement.java]: Changed

File: data-protocols/ids/ids-api-multipart-endpoint-v1/src/main/java/org/eclipse/dataspaceconnector/ids/api/multipart/handler/ArtifactRequestHandler.java [hunk 3]
```diff
@@ -118,11 +120,11 @@ public boolean canHandle(@NotNull MultipartRequest multipartRequest) {
                 .id(contractDefinition.get().getId() + ":" + UUID.randomUUID())
                 .asset(Asset.Builder.newInstance().id(artifactIdsId.getValue()).build())
                 .policy(contractDefinition.get().getContractPolicy())
-                .contractEndDate(Instant.now().getEpochSecond() + 60 * 5 /* Five Minutes */)
-                .contractSigningDate(Instant.now().getEpochSecond() - 60 * 5 /* Five Minutes */)
-                .contractStartDate(Instant.now().getEpochSecond() - 60 * 5 /* Five Minutes */)
-                .consumerAgentId("https://example.com")
-                .providerAgentId("https://example.com")
+                .contractEndDate(ZonedDateTime.ofInstant(Instant.ofEpochMilli(Instant.now().getEpochSecond() + 60 * 5), ZoneId.of("UTC")) /* Five Minutes */)
```
[the remark this answers is not in the document]
Comment (juliapampus) [reply data-protocols/ids/ids-api-multipart-endpoint-v1/src/main/java/org/eclipse/dataspaceconnector/ids/api/multipart/handler/ArtifactRequestHandler.java]: Done

File: spi/src/main/java/org/eclipse/dataspaceconnector/spi/types/domain/contract/agreement/ContractAgreement.java [hunk 4]
```diff
@@ -85,7 +87,7 @@ public String getId() {
      * @return provider id
      */
     @NotNull
-    public String getProviderAgentId() {
+    public URI getProviderAgentId() {
```
[the remark this answers is not in the document]
Comment (juliapampus) [reply spi/src/main/java/org/eclipse/dataspaceconnector/spi/types/domain/contract/agreement/ContractAgreement.java]: Done

File: data-protocols/ids/ids-api-multipart-endpoint-v1/src/main/java/org/eclipse/dataspaceconnector/ids/api/multipart/handler/ContractOfferHandler.java [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+        Objects.requireNonNull(multipartRequest);
+        Objects.requireNonNull(verificationResult);
+
+        var message = (ContractOfferMessage) multipartRequest.getHeader();
+
+        ContractOffer contractOffer = null;
+        try {
+            contractOffer = objectMapper.readValue(multipartRequest.getPayload(), ContractOffer.class);
+        } catch (IOException e) {
+            monitor.debug("ContractOfferHandler: Contract Offer is invalid");
```
[the remark this answers is not in the document]
Comment (juliapampus) [reply data-protocols/ids/ids-api-multipart-endpoint-v1/src/main/java/org/eclipse/dataspaceconnector/ids/api/multipart/handler/ContractOfferHandler.java]: Done

File: core/contract/src/main/java/org/eclipse/dataspaceconnector/contract/negotiation/ConsumerContractNegotiationManagerImpl.java [hunk 6]
```diff
@@ -334,11 +335,11 @@ private int approveContractOffers() {
...
             var agreement = ContractAgreement.Builder.newInstance()
                     .id(UUID.randomUUID().toString())
-                    .contractStartDate(lastOffer.getContractStart().toInstant().toEpochMilli())
-                    .contractEndDate(lastOffer.getContractEnd().toInstant().toEpochMilli())
-                    .contractSigningDate(Instant.now().toEpochMilli())
-                    .providerAgentId(lastOffer.getProvider().toString())
-                    .consumerAgentId(lastOffer.getConsumer().toString())
+                    .contractStartDate(lastOffer.getContractStart())
+                    .contractEndDate(lastOffer.getContractEnd())
+                    .contractSigningDate(ZonedDateTime.now())
```
[the remark this answers is not in the document]
Comment (juliapampus) [reply core/contract/src/main/java/org/eclipse/dataspaceconnector/contract/negotiation/ConsumerContractNegotiationManagerImpl.java]: Created a new issue #341 for that. Mentioned all marked classes.

File: core/contract/src/main/java/org/eclipse/dataspaceconnector/contract/negotiation/ProviderContractNegotiationManagerImpl.java [hunk 7]
```diff
@@ -368,11 +369,11 @@ private int checkConfirming() {
...
                 agreement = ContractAgreement.Builder.newInstance()
                         .id(UUID.randomUUID().toString())
-                        .contractStartDate(lastOffer.getContractStart().toInstant().toEpochMilli())
-                        .contractEndDate(lastOffer.getContractEnd().toInstant().toEpochMilli())
-                        .contractSigningDate(Instant.now().toEpochMilli())
-                        .providerAgentId(lastOffer.getProvider().toString())
-                        .consumerAgentId(lastOffer.getConsumer().toString())
+                        .contractStartDate(lastOffer.getContractStart())
+                        .contractEndDate(lastOffer.getContractEnd())
+                        .contractSigningDate(ZonedDateTime.now())
```
[the remark this answers is not in the document]
Comment (juliapampus) [reply core/contract/src/main/java/org/eclipse/dataspaceconnector/contract/negotiation/ProviderContractNegotiationManagerImpl.java]: see above

File: core/contract/src/main/java/org/eclipse/dataspaceconnector/contract/validation/ContractValidationServiceImpl.java [hunk 8]
```diff
@@ -142,10 +143,10 @@ public boolean validate(ClaimToken token, ContractAgreement agreement, ContractO
     }
 
     private boolean isExpired(ContractAgreement contractAgreement) {
-        return contractAgreement.getContractEndDate() < Instant.now().getEpochSecond();
+        return contractAgreement.getContractEndDate().toInstant().toEpochMilli() < ZonedDateTime.now().toInstant().toEpochMilli();
```
[the remark this answers is not in the document]
Comment (juliapampus) [reply core/contract/src/main/java/org/eclipse/dataspaceconnector/contract/validation/ContractValidationServiceImpl.java]: see above

File: core/contract/src/test/java/org/eclipse/dataspaceconnector/contract/validation/ContractValidationServiceImplTest.java [hunk 9]
```diff
@@ -84,14 +86,14 @@ void verifyContractAgreementValidation() {
...
-                .consumerAgentId("consumer")
+        var agreement = ContractAgreement.Builder.newInstance().id("1")
+                .providerAgentId(URI.create("provider"))
+                .consumerAgentId(URI.create("consumer"))
                 .policy(originalPolicy)
                 .asset(Asset.Builder.newInstance().build())
-                .contractSigningDate(LocalDate.MIN.toEpochDay())
-                .contractStartDate(LocalDate.MIN.toEpochDay())
-                .contractEndDate(LocalDate.MAX.toEpochDay())
+                .contractSigningDate(ZonedDateTime.now())
```
[the remark this answers is not in the document]
Comment (juliapampus) [reply core/contract/src/test/java/org/eclipse/dataspaceconnector/contract/validation/ContractValidationServiceImplTest.java]: see above

**Review by jimmarino — APPROVED** [review]

**Pull request merged**